community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/References.java

```diff
@@ -134,4 +140,79 @@ static boolean hasDirectFlag( long groupReference )
         assert groupReference != NO_ID;
         return (groupReference & DIRECT_FLAG) != 0L;
     }
+
+    /**
+     * Encode a node id as a property reference.
+     *
+     * This allows us to encode a node reference as a property reference so that we can do lookups of changes in the
+     * transaction state.
+     * @param reference the reference to encode
+     * @return The encoded reference
+     */
+    static long setNodeFlag( long reference )
+    {
+        return reference | NODE_MARKER | FLAG_MARKER;
+    }
+
+    /**
+     * Checks if the property reference really is a node reference
+     *
+     * @param nodeReference The reference to check
+     * @return <tt>true</tt>if the reference refers to a node otherwise <tt>false</tt>
+     */
+    static boolean hasNodeFlag( long nodeReference )
+    {
+        assert nodeReference != NO_ID;
+        return (nodeReference & NODE_MARKER) != 0L;
+    }
+
+    /**
+     * Encode a relationship id as a property reference.
+     *
+     * This allows us to encode a relationship reference as a property reference so that we can do lookups of changes in the
+     * transaction state.
+     * @param relationshipReference the reference to encode
+     * @return The encoded reference
+     */
+    static long setRelationshipFlag( long relationshipReference )
+    {
+        return relationshipReference | RELATIONSHIP_MARKER | FLAG_MARKER;
+    }
+
+    /**
+     * Checks if the property reference really is a relationship reference
+     *
+     * @param reference The reference to check
+     * @return <tt>true</tt>if the reference refers to a relationship otherwise <tt>false</tt>
+     */
+    static boolean hasRelationshipFlag( long reference )
+    {
+        assert reference != NO_ID;
+        return (reference & RELATIONSHIP_MARKER) != 0L;
+    }
+
+    /**
+     * Marks the property reference as having changes in the transaction state.
+     * <p>
+     * Setting this flag tells us that the value on disk is not the only source of truth, the transaction state must
+     * also be checked.
+     *
+     * @param reference The reference to set as having tx state
+     * @return The encoded reference.
+     */
+    static long setTxStateFlag( long reference )
+    {
+        return reference | STATE_MARKER | FLAG_MARKER;
+    }
+
+    /**
+     * Checks if the property reference has been marked as having changes in the transaction state.
+     * @param reference The reference to check
+     * @return <tt>true</tt>if the reference has been marked as having changes otherwise <tt>false</tt>
+     */
+    static boolean hasTxStateFlag( long reference )
```

Comment:
This `reference` parameter not updated yet.

Reply:
but in this case it could be either a node a relationship